src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -106,28 +171,43 @@ private static async Task<IEnumerable<IMeasurementGroup>> ParseAsync(Stream data
             return measurementGroups;
         }
 
-        private static IEnumerable<IMeasurementGroup> ParseEventData(IEnumerable<IEventMessage> data, ITelemetryLogger log)
+        private static Tuple<IEnumerable<IMeasurementGroup>, Dictionary<IMeasurement, IEventMessage>> ParseEventData(IEnumerable<IEventMessage> data, ITelemetryLogger logger)
         {
             var partitionId = data.FirstOrDefault()?.PartitionId;
 
-            // Deserialize events into measurements and then group according to the device, type, and other factors
-            return data.Select(e => JsonConvert.DeserializeObject<Measurement>(System.Text.Encoding.Default.GetString(e.Body.ToArray())))
-                .GroupBy(m => $"{m.DeviceId}-{m.Type}-{m.PatientId}-{m.EncounterId}-{m.CorrelationId}")
-                .Select(g =>
+            var dictionary = new Dictionary<IMeasurement, IEventMessage>();
+
+            IEnumerable<IMeasurementGroup> measurementGroups = data.Select(e =>
+            {
+                try
                 {
-                    var measurements = g.ToList();
-                    _ = CalculateMetricsAsync(measurements, log, partitionId).ConfigureAwait(false);
-                    return new MeasurementGroup
-                    {
-                        Data = measurements,
-                        MeasureType = measurements[0].Type,
-                        CorrelationId = measurements[0].CorrelationId,
-                        DeviceId = measurements[0].DeviceId,
-                        EncounterId = measurements[0].EncounterId,
-                        PatientId = measurements[0].PatientId,
-                    };
-                })
-                .ToArray();
+                    var measurement = JsonConvert.DeserializeObject<Measurement>(System.Text.Encoding.Default.GetString(e.Body.ToArray()));

```

Comment:
Suggestion:  I realize this old code, but rather than convert to an array, it would be better to load from a memory stream to avoid excessive memory allocation. See EventMessageJObjectConverter for an example.